openfpga/src/fpga_bitstream/write_text_fabric_bitstream.cpp

```diff
@@ -245,6 +245,142 @@ static int write_memory_bank_flatten_fabric_bitstream_to_text_file(
+ *   manner. Using map key as ordering cannot guarantee the determinstic
+ *
+ * This new way of writting fabric guarantee the WL order in 100% deterministc
+ *   way: either incremental (default) or decremental
+ *
+ * Return:
+ *  - 0 if succeed
+ *  - 1 if critical errors occured
+ *******************************************************************/
+static int fast_write_memory_bank_flatten_fabric_bitstream_to_text_file(
+  std::fstream& fp, const bool& fast_configuration,
+  const bool& bit_value_to_skip, const FabricBitstream& fabric_bitstream,
+  const bool& keep_dont_care_bits, const bool& wl_incremental_order) {
+  int status = 0;
+
+  std::string dont_care_bit = "0";
+  if (keep_dont_care_bits) {
+    dont_care_bit = "x";
+  }
+  const FabricBitstreamMemoryBank* memory_bank =
+    fabric_bitstream.memory_bank_info();
+
+  // Must call this to prepare wls_to_skip
+  (const_cast<FabricBitstreamMemoryBank*>(memory_bank))
+    ->fast_configuration(fast_configuration, bit_value_to_skip);
+
+  fabric_size_t lontest_effective_wl_addr_size =
+    memory_bank->get_lontest_effective_wl_addr_size();
+  /* Output information about how to intepret the bitstream */
+  fp << "// Bitstream length: " << lontest_effective_wl_addr_size << std::endl;
+  fp << "// Bitstream width (LSB -> MSB): ";
+  fp << "<bl_address " << memory_bank->get_total_bl_addr_size() << " bits>";
+  fp << "<wl_address " << memory_bank->get_total_wl_addr_size() << " bits>";
+  fp << std::endl;
+
+  std::vector<fabric_size_t> wl_indexes;
+  for (size_t region = 0; region < memory_bank->datas.size(); region++) {
+    if (wl_incremental_order) {
+      wl_indexes.push_back(0);
+    } else {
+      wl_indexes.push_back(
+        (fabric_size_t)(memory_bank->datas[region].size() - 1));
+    }
+  }
+  for (size_t wl_index = 0; wl_index < lontest_effective_wl_addr_size;
+       wl_index++) {
+    /* Write BL address code */
+    /* cascade region 0, 1, 2, 3 ... */
+    for (size_t region = 0; region < memory_bank->datas.size(); region++) {
+      const fabric_blwl_length& lengths = memory_bank->blwl_lengths[region];
+      fabric_size_t current_wl = wl_indexes[region];
+      while (std::find(memory_bank->wls_to_skip[region].begin(),
+                       memory_bank->wls_to_skip[region].end(),
+                       current_wl) != memory_bank->wls_to_skip[region].end()) {
+        // We would like to skip this
+        if (wl_incremental_order) {
+          wl_indexes[region]++;
+        } else {
+          wl_indexes[region]--;
+        }
+        current_wl = wl_indexes[region];
+      }
+      if (current_wl < memory_bank->datas[region].size()) {
+        const std::vector<uint8_t>& data =
+          memory_bank->datas[region][current_wl];
+        const std::vector<uint8_t>& mask =
+          memory_bank->masks[region][current_wl];
+        for (size_t bl = 0; bl < lengths.bl; bl++) {
+          if (mask[bl >> 3] & (1 << (bl & 7))) {
```

Comment:
These lines are not easy to understand for junior engineers at a glance. Suggest to leave sufficient comments. Also I believe it can be an method, rather than being flatten here?

Reply:
Address via https://github.com/chungshien/OpenFPGA/commit/838cf0d818166d18a51c67d3fc605b990a38e570